### PR TITLE
Unit test imaging subgraphs from schema-test-data (#3873)

### PR DESCRIPTION
**Pull request by danielsotirhos**
https://github.com/DataBiosphere/azul/issues/3873

### Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (freebies & chains)

- [x] Freebies are blocked on this PR                               <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles                      <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain                <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                          <sub>or this PR does not require announcement</sub>
- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

## Author (hotfixes)

- [x] Added `h` tag to commit title and PR targets `prod`           <sub>or this PR does not include a temporary hotfix</sub>
- [x] Added `H` tag to commit title and PR targets `develop`        <sub>or this PR does not include a permanent hotfix</sub>
- [x] Added `hotfix` label to PR                                    <sub>or this PR does not include a hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or this PR does not permanently fix a temporary hotfix</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (before every review)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Primary reviewer (after approval)

- [x] Commented in issue about demo expectations                    <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

### Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                         <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                               <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                        <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate               <sub>or this PR is authored by lead</sub>
- [ ] Pushed merge commit to Gitlab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Build passes on Gitlab
- [ ] Moved issues to `prod` or `Merged prod`                       <sub>or this PR does not represent a promotion</sub>

### Operator (reindex) 

- [ ] Started reindex in target deployment                          <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>
- [ ] Filed backport PR                                             <sub>or this PR does not represent a hotfix</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting issue

### CODE REVIEW

File: test/service/test_response.py [hunk 1]
```diff
@@ -2599,9 +2605,9 @@ def test_summary_cell_counts(self):
         response.raise_for_status()
         summary = response.json()
         self.assertEqual(summary['projectCount'], 1)
-        self.assertEqual(summary['fileCount'], 1 + 1)
+        self.assertEqual(summary['fileCount'], 7)
         self.assertEqual(summary['projectEstimatedCellCount'], 10000.0)
-        self.assertEqual(summary['totalCellCount'], 20000.0 + 20000.0)  # cell suspensions
+        self.assertEqual(summary['totalCellCount'], 20000.0 + 20000.0)
         self.assertEqual(summary['cellCountSummaries'], [
```
Comment (achave11) [inline test/service/test_response.py:2611]: I think the convention for assertions in test is expected followed by actual.

**Review by achave11 — CHANGES_REQUESTED** [review]
Minor changes for adopting assert convention.

**Review by achave11 — APPROVED** [review]
Approved! 👍🏽

File: scripts/can_bundle.py [hunk 2]
```diff
@@ -38,6 +38,8 @@
 
 logger = logging.getLogger(__name__)
 
+logging.getLogger("humancellatlas.data.metadata.helpers.staging_area").setLevel(logging.DEBUG)
```
Comment (hannes-ucsc) [inline scripts/can_bundle.py:41]: No side-effects at module level.

File: test/service/test_response.py [hunk 3]
```diff
@@ -2599,9 +2605,9 @@ def test_summary_cell_counts(self):
         response.raise_for_status()
         summary = response.json()
         self.assertEqual(summary['projectCount'], 1)
```
Comment (hannes-ucsc) [inline test/service/test_response.py]: Can we assert assay type?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
The last commit should be the first ideally.

File: test/service/data/pfb_manifest.results.json [hunk 4]
```diff
@@ -3040,7 +3040,7 @@
                 ""
             ],
             "submission_date": "2021-01-01T00:00:00.000000Z",
-            "update_date": ""
+            "update_date": "2021-01-01T00:00:00.000000Z"
```
[the remark this answers is not in the document]
Comment (danielsotirhos) [reply test/service/data/pfb_manifest.results.json:3043]: The test case `test_pfb_manifest` makes use of the canned bundle `4da04038-adab-59a9-b6c4-3a61242cc972`, which is a bundle updated by this PR.

File: test/service/test_response.py (hunk 3 above)
Comment (danielsotirhos) [reply test/service/test_response.py]: I've added a test case for the protocol values, however we currently suffer from a lack of test data.

Not only did the recent schema changes rename `imaging_protocol`'s `target`  to `probe`, it also made `probe` [optional](https://schema.staging.data.humancellatlas.org/type/protocol/imaging/11.3.0/imaging_protocol). The bundles from the canned staging area do have an [imaging protocol](https://github.com/HumanCellAtlas/schema-test-data/blob/master/tests/metadata/imaging_protocol/50150415-898c-5dff-826e-55f9cb937b93_2021-01-01T00:00:00.000000Z.json) but do not include a `probe` (or `target`) and thus no `assay type`.

The test case I added asserts the `protocols` inner entity values, which for `assayType` is `[]`.

File: test/service/test_response.py [hunk 5]
```diff
@@ -2827,10 +2827,16 @@ class TestSchemaTestDataCannedBundle(WebServiceTestCase):
     @classmethod
     def bundles(cls) -> List[BundleFQID]:
         return [
-            # Bundles from the canned staging area, both for project 90bf705c
+            # Bundles for project 90bf705c from commit 7a03924481
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2830]: That can't be right. That commit precedes the merging of https://github.com/HumanCellAtlas/schema-test-data/pull/16

File: test/service/test_response.py [hunk 6]
```diff
@@ -2827,10 +2827,16 @@ class TestSchemaTestDataCannedBundle(WebServiceTestCase):
...
             # https://github.com/HumanCellAtlas/schema-test-data/
```
Comment (hannes-ucsc) [inline test/service/test_response.py:2831]:
```suggestion
            # https://github.com/HumanCellAtlas/schema-test-data/commit/7a03924481
```

File: test/service/test_response.py (hunk 5 above)
Comment (hannes-ucsc) [reply test/service/test_response.py:2830]: Now you've gotta somehow prove to me that the cans you committed come from that commit, or from a commit, whatever the right commit is.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/service/test_response.py (hunk 5 above)
Comment (danielsotirhos) [reply test/service/test_response.py:2830]: @hannes-ucsc : "I misinterpreted the history depicted in the screenshot. In reality commit 7a03924481 is the most recent commit on the PR branch, and the subsequent merge should not affect the cans. However, I still think we should re-can the merge commit and have the link in the comment above the test refer to the merge commit as well. Re-canning should not modify the cans."

**Comment by danielsotirhos** [comment]
Verified freshly canned bundles from the merge commit [2a62a7f49270d1e95e65a](https://github.com/HumanCellAtlas/schema-test-data/commit/2a62a7f49270d1e95e65a3cf8c788b8700d2e543) match the existing canned bundles.
[terminal-log.txt](https://github.com/DataBiosphere/azul/files/8257818/terminal-log.txt)

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Debumped because https://github.com/DataBiosphere/azul/pull/3899#discussion_r827356292

**Pull request merged**